### PR TITLE
support concurent remote connetion

### LINKED ISSUE #18: tcp router client: performance issue
The current implementation of the tcp router client gives really poor performance.
We should investigate and see why we get such a small throughput

### PR DESCRIPTION
**Pull request by zaibon**
the client now open a multiplexed connection to the server,
does the handshake then start listening for incoming connection from the
server.

This allow to received multiple incoming connection concurently
and handle them properly. Everytime the client receive a new connection
it open one to the local app too.

fix #18

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -1,58 +1,90 @@
...
 	// secret used to identify the connection in the tcp router server
 	secret []byte
+
+	// connection to the tcp router server
+	remoteSession *yamux.Session
 }
 
 // NewClient creates a new TCP router client
-func NewClient(secret string) *Client {
+func NewClient(secret string, local, remote string) *Client {
```
Comment (xmonader) [inline client.go]: should be secret, local, remote string

File: client.go [hunk 2]
```diff
@@ -1,58 +1,90 @@
...
 	}
 }
 
-func (c *Client) ConnectRemote(addr string) error {
+func (c Client) Start(ctx context.Context) error {
+	if err := c.connectRemote(c.remoteAddr); err != nil {
+		return fmt.Errorf("failed to connect to TCP router server: %w", err)
+	}
+
+	log.Info().Msg("start hanshake")
```
Comment (xmonader) [inline client.go]: typo "handshake"

File: client.go [hunk 3]
```diff
@@ -1,58 +1,90 @@
...
+	if err := c.handshake(); err != nil {
+		return fmt.Errorf("failed to hanshake with TCP router server: %w", err)
+	}
+	log.Info().Msg("hanshake done")
```
Comment (xmonader) [inline client.go]: typo "handshake"

File: client.go [hunk 4]
```diff
@@ -1,58 +1,90 @@
...
+	return c.listen(ctx)
+}
+
+func (c *Client) connectRemote(addr string) error {
 	if len(c.secret) == 0 {
 		return fmt.Errorf("no secret configured")
 	}
 
-	conn, err := net.Dial("tcp", addr)
+	tcpAddr, err := net.ResolveTCPAddr("tcp", addr)
```
Comment (xmonader) [inline client.go:53]: this returns only 1 addr even if we have 3 tcprouters on specific domain no?  also requires a port which we should ignore if not specified to be 18000

File: server.go [hunk 5]
```diff
@@ -35,15 +36,16 @@ func (o ServerOptions) ClientsAddr() string {
 }
 
 type Server struct {
-	ServerOptions     ServerOptions
-	DbStore           store.Store
-	Services          map[string]Service
-	backendM          sync.RWMutex
```
Comment (xmonader) [inline server.go:41]: probably we missed this, it was meant for reloading from routerconfig.toml when sighup is sent to the server to update services from the toml file (maybe to be revisited later)

File: client.go (hunk 4 above)
Comment (zaibon) [reply client.go:53]: a client always connect to a single remote router. If you need to connect to multiple router then you instantiate multiple client. Check the client main function to see how this works.
What's up with the port 18000 ?

File: server.go (hunk 5 above)
Comment (zaibon) [reply server.go:41]: yes this signal management is nowhere. Lets leave this up to be implemented in another PR.

**Review by LeeSmet — APPROVED** [review]

**Pull request merged**